### PR TITLE
After review

### PR DESCRIPTION
**Pull request by BarTes8**

### CODE REVIEW

File: shm/inc/Cargo.hpp [hunk 1]
```diff
@@ -2,17 +2,19 @@
 
 #include <string>
 
+//class resposible for manage commodity in game
 class Cargo {
 public:
-    Cargo(const std::string&, size_t, size_t);
+    //constructs an object of cargo in game
```
Comment (MoreThanADC) [inline shm/inc/Cargo.hpp:10]: Im not sure about this comment, probably everyone knows what the constructor is for :)

**Review by MoreThanADC — APPROVED** [review]
Well done

File: shm/inc/Island.hpp [hunk 2]
```diff
@@ -2,21 +2,23 @@
...
-        Coordinates(size_t, size_t);
-        bool operator==(const Coordinates&) const;
+        Coordinates(size_t positionX, size_t positionY);
+        bool operator==(const Coordinates& position) const;
     private:
         const size_t positionX_ = 0;
         const size_t positionY_ = 0;
     };
 
+    Island(const Island::Coordinates& position);
```
Comment (mariuszlisowski) [inline shm/inc/Island.hpp]: The constructor should be marked as 'explicit'.

File: shm/inc/Player.hpp [hunk 3]
```diff
@@ -7,9 +7,10 @@
 #include <memory>
 #include <numeric>
 
+//class responsible for manage user in game
```
Comment (mariuszlisowski) [inline shm/inc/Player.hpp]: 1. Redundant headers  \<algorithm\> &  \<numeric\>  - the second one should be in source file (if std::accumulate used).

2. Move include of Cargo and Ship class to source file. 
Instead use forward declaration of needed classes as follows:

#include \<cstddef\>    // for size_t
#include \<memory\>

#include "shm/inc/Delegate.hpp"

class Ship;
class Cargo;

File: shm/src/Player.cpp [hunk 4]
```diff
@@ -9,15 +9,12 @@ size_t Player::getSpeed() const {
 }
   
 std::shared_ptr<Cargo> Player::getCargo(size_t index) const {
-    return ship_->getCargo(index);
+    if (ship_) {
+        return ship->getCargo(index);
```
Comment (mariuszlisowski) [inline shm/src/Player.cpp]: ship_->

File: shm/src/Player.cpp [hunk 5]
```diff
@@ -9,15 +9,12 @@ size_t Player::getSpeed() const {
...
 
 size_t Player::countAvailableSpace() const {
-    return ship_->getCapacity() - 
-            std::accumulate(ship_->getVectorCargo().cbegin(),
-                    ship_->getVectorCargo().cend(),
-                    static_cast<size_t>(0),
-                    [](size_t sum, const auto& cargo) {
-                        return sum += cargo->getAmount();
-                    });
+    size_t availableSpace;
```
Comment (mariuszlisowski) [inline shm/src/Player.cpp]: I'll find the solution soon

File: shm/src/Cargo.cpp [hunk 6]
```diff
@@ -9,16 +11,20 @@ Cargo::Cargo(const std::string& name, size_t amount, size_t basePrice)
 {}
 
 Cargo& Cargo::operator+=(const size_t amount) {
+    if (amount_ + amount > std::numeric_limits<size_t>::max()) {
+        std::cerr << "You don't have enough space on the ship\0";
+        return *this;
+    }
```
Comment (mariuszlisowski) [inline shm/src/Cargo.cpp:15]: else

File: shm/src/Cargo.cpp [hunk 7]
```diff
@@ -9,16 +11,20 @@ Cargo::Cargo(const std::string& name, size_t amount, size_t basePrice)
 {}
 
 Cargo& Cargo::operator+=(const size_t amount) {
+    if (amount_ + amount > std::numeric_limits<size_t>::max()) {
+        std::cerr << "You don't have enough space on the ship\0";
+        return *this;
```
Comment (mariuszlisowski) [inline shm/src/Cargo.cpp]: redundant

File: shm/src/Cargo.cpp [hunk 8]
```diff
@@ -9,16 +11,20 @@ Cargo::Cargo(const std::string& name, size_t amount, size_t basePrice)
 {}
 
 Cargo& Cargo::operator+=(const size_t amount) {
+    if (amount_ + amount > std::numeric_limits<size_t>::max()) {
+        std::cerr << "You don't have enough space on the ship\0";
```
Comment (mariuszlisowski) [inline shm/src/Cargo.cpp]: don't handle error case here

File: shm/src/Cargo.cpp [hunk 9]
```diff
@@ -9,16 +11,20 @@ Cargo::Cargo(const std::string& name, size_t amount, size_t basePrice)
 {}
 
 Cargo& Cargo::operator+=(const size_t amount) {
+    if (amount_ + amount > std::numeric_limits<size_t>::max()) {
```
Comment (mariuszlisowski) [inline shm/src/Cargo.cpp]: A huge number of cargo! do we need so many? The idea was to return ::max() to inform about an error case - not to compare.

File: shm/src/Ship.cpp [hunk 10]
```diff
@@ -14,21 +14,21 @@ Ship::Ship(int id, size_t speed, size_t maxCrew)
...
 {}
 
-Ship& Ship::operator+=(const size_t amount) {
-    if (amount + crew_ <= maxCrew_) {
-        crew_ += amount;
-    } else {
-        std::cerr << "Maximum amount of crew is " << maxCrew_ << '\n';
+Ship& Ship::operator+=(const size_t crew) {
+    if (crew_ + crew > maxCrew_) {
+        std::cerr << "Too many sailors\n";
```
Comment (mariuszlisowski) [inline shm/src/Ship.cpp]: output not here, set crew_ to ::max() or throw an exception and handle it outside the method

File: shm/src/Ship.cpp [hunk 11]
```diff
@@ -14,21 +14,21 @@ Ship::Ship(int id, size_t speed, size_t maxCrew)
...
+        return *this;
```
Comment (mariuszlisowski) [inline shm/src/Ship.cpp]: redundant

File: shm/src/Ship.cpp [hunk 12]
```diff
@@ -14,21 +14,21 @@ Ship::Ship(int id, size_t speed, size_t maxCrew)
...
     }
```
Comment (mariuszlisowski) [inline shm/src/Ship.cpp:22]: else

File: shm/src/Ship.cpp [hunk 13]
```diff
@@ -14,21 +14,21 @@ Ship::Ship(int id, size_t speed, size_t maxCrew)
...
 }
 
-Ship& Ship::operator-=(const size_t amount) {
-    if (crew_ <= amount) {
-        crew_ -= amount;
-    } else {
-        std::cerr << "Amount of crew can't be under 0\n";
+Ship& Ship::operator-=(const size_t crew) {
+    if (crew_ < crew) {
+        std::cerr << "Number of sailors lower than 0\n";
```
Comment (mariuszlisowski) [inline shm/src/Ship.cpp]: output not here, set crew_ to ::max() or throw an exception and handle it outside the method

File: shm/src/Ship.cpp [hunk 14]
```diff
@@ -14,21 +14,21 @@ Ship::Ship(int id, size_t speed, size_t maxCrew)
...
+        return *this;
     }
```
Comment (mariuszlisowski) [inline shm/src/Ship.cpp:30]: else

File: shm/src/Ship.cpp [hunk 15]
```diff
@@ -14,21 +14,21 @@ Ship::Ship(int id, size_t speed, size_t maxCrew)
...
 
-Ship& Ship::operator-=(const size_t amount) {
-    if (crew_ <= amount) {
-        crew_ -= amount;
-    } else {
-        std::cerr << "Amount of crew can't be under 0\n";
+Ship& Ship::operator-=(const size_t crew) {
+    if (crew_ < crew) {
+        std::cerr << "Number of sailors lower than 0\n";
+        return *this;
```
Comment (mariuszlisowski) [inline shm/src/Ship.cpp]: redundant

File: shm/src/Ship.cpp [hunk 16]
```diff
@@ -37,8 +37,13 @@ void Ship::setName(const std::string& name) {
 }
 
 std::shared_ptr<Cargo> Ship::getCargo(size_t index) const {
-    if (cargo_.size() <= index) {
+    if (index>= cargos_.size()) {
+        std::cerr << "Invalid cargo\n";
```
Comment (mariuszlisowski) [inline shm/src/Ship.cpp]: output not here

**Review by mariuszlisowski — CHANGES_REQUESTED** [review]

File: shm/inc/Cargo.hpp (hunk 1 above)
Comment (BarTes8) [reply shm/inc/Cargo.hpp:10]: yes, i forgot to delet it :-)

File: shm/src/Ship.cpp (hunk 16 above)
Comment (BarTes8) [reply shm/src/Ship.cpp]: this is anyway the function we need to rewrite if you see the review, but i don't know how

**Review by mariuszlisowski — APPROVED** [review]

**Pull request merged**